File: Task41/app/src/main/java/edu/deakin/s600152989/sit305/task41/TaskViewModel.java
package edu.deakin.s600152989.sit305.task41;

import android.app.Application;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import java.util.List;

import androidx.lifecycle.MutableLiveData;

public class TaskViewModel extends AndroidViewModel {

    private TaskRepository repository;
    private LiveData<List<Task>> allTasks;
    private MutableLiveData<String> errorMessage = new MutableLiveData<>();

    public TaskViewModel(Application application) {
        super(application);
        repository = new TaskRepository(application);
        allTasks = repository.getAllTasks();
    }

    public LiveData<List<Task>> getAllTasks() {
        return allTasks;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public void insert(Task task) {
        try {
            repository.insert(task);
        } catch (Exception e) {
            errorMessage.postValue("Error inserting task: " + e.getMessage());
        }
    }

    public void update(Task task) {
        try {
            repository.update(task);
        } catch (Exception e) {
            errorMessage.postValue("Error updating task: " + e.getMessage());
        }
    }

    public void delete(Task task) {
        try {
            repository.delete(task);
        } catch (Exception e) {
            errorMessage.postValue("Error deleting task: " + e.getMessage());
        }
    }

    public void deleteAllTasks() {
        try {
            repository.deleteAll();
        } catch (Exception e) {
            errorMessage.postValue("Error deleting all tasks: " + e.getMessage());
        }
    }
}

File: Task41/app/src/main/java/edu/deakin/s600152989/sit305/task41/TaskViewModel.kt
package edu.deakin.s600152989.sit305.task41

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData

class TaskViewModel(application: Application) : AndroidViewModel(application) {
    private val repository =
        TaskRepository(application)
    val allTasks: LiveData<List<Task?>?>? =
        repository.allTasks
    private val errorMessage = MutableLiveData<String?>()

    fun getErrorMessage(): LiveData<String?> {
        return errorMessage
    }

    fun insert(task: Task?) {
        try {
            repository.insert(task)
        } catch (e: Exception) {
            errorMessage.postValue("Error inserting task: " + e.message)
        }
    }

    fun update(task: Task?) {
        try {
            repository.update(task)
        } catch (e: Exception) {
            errorMessage.postValue("Error updating task: " + e.message)
        }
    }

    fun delete(task: Task?) {
        try {
            repository.delete(task)
        } catch (e: Exception) {
            errorMessage.postValue("Error deleting task: " + e.message)
        }
    }

    fun deleteAllTasks() {
        try {
            repository.deleteAll()
        } catch (e: Exception) {
            errorMessage.postValue("Error deleting all tasks: " + e.message)
        }
    }
}
